nextjsUIapp/src/app/knowledge-base/[slug]/[id]/page.tsx

```diff
@@ -1,122 +1,104 @@
+            if (response.status === "success" && response.message?.results?.bindings) {
+                return response.message.results.bindings;
             }
-        } catch (e) {
-            const error = e as Error;
-
-            setError(error.message);
-        } finally {
-            setLoading(false);
+            return [];
+        } catch (err) {
+            console.error("Error fetching data:", err);
+            return [];
         }
-
     };
 
     const loadData = async () => {
-        setMainCardTitle("");
-        setData([]);
-        setMainCardDescription("");
-
         try {
-            // Fetch the mapper file
             const page = enititycardmapperconfig.EntityViewCardsMaper.find((page) => page.slug === slug);
             const filename = page ? page.filename : "";
 
-            // Read entity card design models dynamically
             const model_data = await import(`@/src/app/components/${filename}`);
             const extracted_data = model_data.default;
 
-            // Set card title and description
             setMainCardTitle(extracted_data?.name || "");
             setMainCardDescription(extracted_data?.description || "");
             setExtractedBoxes(extracted_data?.boxes || []);
-            console.log(extracted_data?.boxes);
         } catch (error) {
-            console.error('Failed to fetch YAML data or Query:', error);
+            console.error("Failed to fetch YAML data:", error);
         }
     };
 
     useEffect(() => {
         loadData();
-    }, []);
+    }, [slug]);
 
+    useEffect(() => {
+        const fetchBoxData = async () => {
+            for (const box of extractedBoxes) {
+                if (box.cardtype === "card" && box.sparql_query) {
+                    const query = box.sparql_query.replace(/\{0\}/g, decodeURIComponent(id));
+                    const boxData = await fetchData({ sparql_query: query });
+                    const cleanedData = await extractPredicateObjectPairs(boxData);
+                    const formattedData = await formatextractPredicateObjectPairs(cleanedData);
+                    setData((prevData) => ({
+                        ...prevData,
+                        [box.slug]: formattedData,
+                    }));
+                }
+            }
+        };
+
+        if (extractedBoxes.length > 0) {
+            fetchBoxData();
+        }
+    }, [extractedBoxes, id]);
 
     return (
         <div className="kb-page-margin">
-            <SideBarKBFromConfig/>
+            <SideBarKBFromConfig />
             <div className="grid fix-left-margin grid-cols-1">
                 <div className="w-full bg-white shadow-md rounded-lg overflow-hidden">
                     <div className="p-4">
-                        <h2 className="text-xl font-bold">{mainCardTitle}</h2>
+                        <h2 className="text-xl font-bold"> {decodeURIComponent(id).substring(decodeURIComponent(id).lastIndexOf("/") + 1)}</h2>
```

Comment:
Perhaps move to a function rather than inline?

Reply:
If later this needs to be re-used in multiple occurrence then I would update it to functions.